[4-9단계 - 방탈출 예약 관리] 리원(김지윤) 미션 제출합니다.

awrion3 opened the pull request:
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr/studylogs/4192)에 셀프 체크를 작성했나요?

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

- 웹 요청과 응답에 따른 단계별 요구사항을 준수하였습니다.
- 도메인, 리포지토리, 서비스, 컨트롤러의 역할을 분리해 계층형 아키텍쳐를 구성하고자 했습니다.
- 널 값 입력이나 중복된 값, 존재하지 않는 아이디 값에 대한 기초적인 예외 처리를 구현했습니다.
- 각 객체가 하나의 역할을 지닐 수 있도록 배분하고자 노력했습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [x] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?

```
안녕하세요, 제이미! 리뷰이 리원입니다. 다시 만나서 반갑습니다. 🙇🏻
이전에 스텝 1에서 말씀하신 피드백을 참고하여 다음과 같은 부분을 수정했습니다.
- 상태 코드를 ok로 일관되게 유지함에 따라 Response Entity로 감싸서 반환하는 대신 RestController를 사용해 각 객체를 반환하는 방식을 채택했습니다.
- 일부 상수명과 관련해 의미가 모호하거나 boolean 등의 타입으로 오해할 수 있는 부분에 대해 수정했습니다.
- 단계별 요구사항과 관련된 기능 목록 작성에 있어 좀 더 구체적으로 변경 사항을 정리하였습니다.
또한 스텝 2를 진행하면서 고민해 볼 수 있었던 부분은 다음과 같습니다.
- 서비스, 컨트롤러에 대한 의존성 주입과 관련하여 객체가 생성되는 시점에 주입하여 불변성을 보장할 수 있다는 점에서 생성자를 통한 주입을 선택했습니다.
- 계층형 아키텍쳐를 적용하는 과정에서 service를 도입했지만, 현재로서는 repository와 비교했을 때 유의미한 차이가 없다고 느껴져 test를 생략했습니다.
이번에도 시간 내어 리뷰해주셔서 감사합니다. ☺️
```

### 📌 코드리뷰 질문 그라운드룰
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

awrion3 committed:
feat: add admin controller for getting main page

awrion3 committed:
feat: add read reservation

awrion3 committed:
feat: add create and delete reservation

awrion3 committed:
refactor: use dto and add reservation copy method

awrion3 committed:
refactor: use record for reservation

awrion3 committed:
docs: add simple feature list

awrion3 committed:
refactor: use require non null for reservation parameter

awrion3 committed:
docs: update feature list

awrion3 committed:
refactor(Reservation): remove copy method and use exception header

- add null value exception test

awrion3 committed:
feat(Reservations): add find, save, delete by id methods

- add duplication check and test cases

awrion3 committed:
feat: add request and response dto for conversion

awrion3 committed:
refactor: split controllers and test cases

- add controllers for displaying pages and handling reservations
- add line break for application class

awrion3 committed:
refactor: use constant variable name for formatter

awrion3 committed:
refactor: use simplified response entity method

awrion3 committed:
refactor: use id class for atomic long

- use constant for controller test
- use request mapping for controller

awrion3 committed:
refactor: use di and final for reservations

- rename constant for exception message
- apply synchronized keyword for reservations

awrion3 committed:
feat: remove old files and add database setups

awrion3 committed:
feat: add reservation and reservation time

- use reservation time instead of local time in reservation class

awrion3 committed:
feat: add request and response dto for reservation time

- change fields of reservation dto

awrion3 committed:
feat: add reservation time repository and test

- change reservation repository methods according to updated reservation class

awrion3 committed:
feat: add reservation and reservation time service

awrion3 committed:
feat: add reservation time controller

- add display time page to display controller
- apply reservation service to reservation controller

awrion3 committed:
docs: update feature list for step 2

awrion3 committed:
Merge branch 'awrion3' into step2

awrion3 committed:
fix: revert changes caused by merging process

jamie9504 commented on README.md:
```diff
@@ -8,19 +8,314 @@
+
+---
+
+### 기능 목록: 7 ~ 9 단계
```
기능 목록을 1~3단계와 4~6단계, 7~9단계를 나눴는데 나눈 이유가 무엇인가요?
`###`레벨만 봤을 때는 기준이 드러나진 않는 것 같아요.

awrion3 replied:
음 네, 미션에 따른 단계별 진행 내용을 정리하고자 기능 목록을 세부적으로 나누었는데요, 
다시 보니 기능 목록을 분리하는 기준이 되기에는 다소 명확하지 않다는 생각도 드네요. 
리팩토링을 통해 하나의 기능 목록만을 두고 관련 내용들을 담을 수 있도록 변경했습니다.

jamie9504 commented on src/main/java/roomescape/controller/AdminDisplayController.java:
```diff
@@ -15,6 +15,11 @@ public String displayAdminMain() {
+
+    @GetMapping("/time")
+    public String displayAdminTime() {
+        return "/admin/time";
```
`/admin/{}` vs `admin/{}` 일관성 있는 응답 포맷이면 좋겠죠?

awrion3 replied:
아 네, 그렇네요! 상대 경로를 따르는 방식으로 일관성 있게 통일하여 리팩토링 했습니다.

jamie9504 commented on src/main/java/roomescape/controller/AdminReservationController.java:
```diff
@@ -0,0 +1,40 @@
+        return ReservationResponse.toResponses(reservationService.findAll());
+    }
+
+    @PostMapping
```
`noContent()와 같은 상태 코드를 2단계에서 도입할지 고민하고 있어서 이렇게 사용해보게 된 것 같습니다.`

미션과 별개로 시간이 있을 때 Get, Post, Delete 등의 Http 요청 메서드와 응답 코드, RESTFUL API에 대한 학습을 해두면 이번 레벨이 한층 수월할 거에요.

awrion3 replied:
네, 말씀해주신 해당 개념들에 대해서 좀 더 학습해보도록 하겠습니다! 피드백 감사합니다. ☺️

jamie9504 commented on src/main/java/roomescape/controller/AdminReservationController.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @PostMapping
+    public Reservation getReservationById(@RequestBody ReservationRequest reservationRequest) {
```
`getReservationById`, `save`는 명백히 다른 행위를 할 것으로 보이는데 적절한 메서드 네이밍인가요?

awrion3 replied:
아 네, 반환되는 값에 집중하다 보니 행위를 잘 반영하지 못한 네이밍을 선택하게 되었네요.
다시 리팩토링 하는 과정에서는 save 등을 사용해서 의도를 좀 더 잘 드러낼 수 있게 변경했습니다.

jamie9504 commented on src/main/java/roomescape/controller/AdminReservationController.java:
```diff
@@ -0,0 +1,40 @@
+
+@RestController
+@RequestMapping("/reservations")
+public class AdminReservationController {
+
+    private final ReservationService reservationService;
+
```
클래스 네이밍에 `Admin`이 붙은 이유는 무엇인가요?

awrion3 replied:
음 네, 예약 및 시간 내역의 조회, 추가, 삭제 작업을 하는 관리자 페이지라는 생각이 들어서 이러한 클래스 네이밍을 선택하게 된 것 같습니다.

jamie9504 replied:
해당 API가 Admin에서만 사용될 수 있는 API라고 생각했다면 문제 없습니다.
그렇지만 그렇다면 Path가 `@RequestMapping("/reservations")`와 같이 설정되지 않았을 수 있겠네요.

awrion3 replied:
오 그렇군요, 말씀하신 부분을 고려해보니 컨트롤러명을 알맞게 대응하여 설정하는 것이 더 낫겠다는 생각이 드네요. 수정해보겠습니다!

jamie9504 commented on src/main/java/roomescape/model/Reservation.java:
```diff
@@ -1,16 +1,41 @@
+    private static final String NULL_VALUE_EXCEPTION = "널 값은 입력될 수 없습니다.";
 
-    public Reservation {
-        Objects.requireNonNull(name, IS_NULL_VALUE);
-        Objects.requireNonNull(date, IS_NULL_VALUE);
-        Objects.requireNonNull(time, IS_NULL_VALUE);
```
date, time에 대한 검증이 제거된 이유는 무엇인가요?

`IS_NULL_VALUE`가 boolean을 반환할 것 같았다면,
`NULL_VALUE_EXCEPTION`은 Exception을 반환할 것 같기도 하네요.

예외 메시지라는 의미를 담거나, 문자열이라는 의미를 담아주세요.

awrion3 replied:
음, date와 time의 경우, 현 시점에서는 사용자가 입력하지 않고 제공된 값을 선택하는 방식이어서 검증을 생략해보게 되었습니다.
또한 상수명과 관련해서 exception만 작성한 경우에도 반환값이 의도와 다르게 전달될 수 있겠네요, 예외 메시지임을 명시하는 방향으로 수정했습니다!

jamie9504 replied:
다른 개발자와 협업을 한다면, Reservation을 사용할 때 현 시점 기반의 validate인지 세세히 파악하고 사용할 수 있을지도 고민해보세요!

생략을 하는 경우도 타인 또는 훗날의 본인이 예상할 수 있는 생략을 하면 좋습니다.

awrion3 replied:
아 네, 생략과 관련해 예상할 수 있어야 한다는 점을 고려해 봐야겠네요! 다음 단계에서는 검증 로직을 좀 더 구현해보도록 하겠습니다.

jamie9504 commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,105 @@
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    private final RowMapper<Reservation> rowMapper = (resultSet, rowNum) -> new Reservation(
```
해당 멤버변수의 선언이 생성자 아래에 위치한 이유는 무엇인가요?

awrion3 replied:
엇 네, rowMapper를 무심코 메서드와 같이 배치해두었네요, 필드 쪽으로 다시 리팩토링 하겠습니다!

jamie9504 replied:
상수로 선언하는 것과 멤버 변수로 선언하는 것은 어떠한 차이를 두고 있나요?

awrion3 replied:
오 네, 레포지토리 객체마다 RowMapper를 다르게 구성할 가능성은 낮으니 상수로 선언하는 것이 더 낫겠군요. 해당 부분 리팩토링 했습니다!

jamie9504 commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,105 @@
+        List<Reservation> reservations = readAllReservations();
+
+        boolean isDuplicate = reservations.stream()
+                .anyMatch(reserve -> reserve.getDate().equals(reservation.getDate()) &&
+                        reserve.getTime().getId().equals(reservation.getTime().getId())
```
디미터의 법칙을 위반하고 있는 것 같은데요.
1레벨에서 배웠던 것을 생각해보고, 전반적으로 리팩터링해볼까요?

awrion3 replied:
오 네, 그렇네요. 해당 부분은 `reservation`과 `reservationTime`에 비교 메서드를 추가하는 방향으로 리팩토링 했습니다.

jamie9504 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,28 @@
+import roomescape.model.Reservation;
+import roomescape.repository.ReservationRepository;
+
+@Service
+public class ReservationService {
```
`계층형 아키텍쳐를 적용하는 과정에서 service를 도입했지만, 현재로서는 repository와 비교했을 때 유의미한 차이가 없다고 느껴져 test를 생략했습니다.`

그렇다면 해당 프로그램에서 Service 레이어의 역할은 무엇인가요?

---

[숙제] 이번 단계에서 도입해본 개념의 키워드들을 쭉 나열해보고, 리원이 생각하는 우선순위대로 정렬해보세요.
그 중 3위까지에 대한 조금 더 깊은 학습을 해보고, 정리해주세요.

awrion3 replied:
음, 현재 시점에서는 서비스 레이어가 예약 혹은 시간 관리와 관련된 리포지토리를 적절히 호출하는 역할만 맡고 있습니다.
다만, 향후 요구사항이 늘어나 여러 도메인 로직을 조율하게 될 경우, 이러한 작업은 서비스 레이어에서 담당하게 되는데요.
이에 대비해서 확장을 위한 기반을 마련하기 위해 해당 계층을 미리 도입해보게 된 것 같습니다.

***

HTTP 요청 메서드와 응답 코드, Spring 어노테이션, Spring MVC, JDBC Template, Layered Architecture 등의 개념을 적용해볼 수 있었던 것 같습니다.

우선 HTTP 요청 메서드는 클라이언트가 서버에 무엇을 하고 싶은지 명확하게 전달하기 위해 필요합니다.
따라서 리소스를 조회, 생성, 수정, 삭제하는 다양한 의도를 구분하여, 서버가 요청을 올바르게 처리할 수 있도록 돕습니다.
- GET: 리소스를 조회합니다.
- HEAD: 본문 없이 리소스 조회(메타데이터만 확인)합니다.
- POST: 새 리소스를 생성하고 서버에 데이터를 제출합니다.
- PUT: 타겟 리소스를 전체 대체합니다.
- PATCH: 타겟 리소스의 일부를 수정합니다.
- DELETE: 특정 리소스를 삭제합니다.
- OPTIONS: 타겟 리소스와 소통 가능한 방법을 확인합니다.
- TRACE: 요청이 서버까지 가는 경로를 테스트합니다.
- CONNECT: 타겟 리소스가 있는 서버와 터널을 생성합니다.

그리고 동일한 요청을 한 번 보내는 것과 여러 번 연속으로 보내는 것이 같은 효과를 지녀 서버 상태도 변하지 않는 경우, 해당 HTTP 메서드는 멱등성을 가진다고 합니다.
- 멱등성 메서드로는 GET, HEAD, PUT, DELETE, OPTIONS, TRACE가 있습니다.
- 비멱등성 메서드로는 POST, PATCH, CONNECT가 있습니다.

POST는 같은 요청을 반복하면 새 리소스가 계속 생성될 수 있어 결과가 달라집니다.
PATCH는 기존 리소스 일부를 수정하거나 추가하는 데 사용되므로, 요청 반복 시 결과가 달라질 수 있습니다.
CONNECT의 경우, 같은 요청을 반복하면 매번 새로운 터널이 생성되어 서버 상태에 영향을 줄 수 있기 때문입니다.

***

또한 스프링에서는 다양한 어노테이션을 제공해 개발자가 코드를 더 간결하고 명확하게 작성할 수 있도록 도와주며, 테스트와 관련해서는 다음과 같은 어노테이션들을 제공하고 있습니다.

- `@SpringBootTest`는 애플리케이션의 모든 객체를 실제 빈으로 등록하여 전체 Spring Boot 애플리케이션을 통합 테스트할 때 사용됩니다. 즉 애플리케이션의 여러 레이어를 실제 환경처럼 로드하여 전체 시스템이 잘 동작하는지 확인할 수 있습니다.
- `@WebMvcTest`는 웹 요청과 관련된 빈들만 테스트할 때 사용됩니다. 컨트롤러 등 웹 레이어만 로드되며, 서비스나 리포지토리 등 다른 컴포넌트는 로드되지 않습니다.
- `@JdbcTest`는 리포지토리 레이어와 관련해 데이터베이스와의 연결 및 SQL 실행과 관련된 빈들을 테스트할 때 사용됩니다. 즉, JdbcTemplate을 주입받아 DB와의 상호작용을 테스트하는 데 사용되며, 쿼리와 관련된 로직을 독립적으로 테스트할 수 있습니다.

***

한편 Spring MVC는 Spring 프레임워크의 웹 애플리케이션 개발을 지원하는 모듈로, Model-View-Controller 아키텍처를 기반으로 클라이언트의 요청을 처리하고 응답을 생성합니다.
즉, HTTP 요청에 대해 DispatcherServlet을 통해서 `@Controller`로 전달되면, 해당 Controller는 Service 계층을 통해 비즈니스 로직을 수행하고, Repository 계층을 통해 데이터베이스와 상호작용하며, 최종적으로 View를 렌더링하여 응답을 반환합니다.

이때 `@Controller`는 웹 요청을 처리하는 컴포넌트임을 나타내는 어노테이션으로, 이와 함께 `@RequestMapping` 등을 활용해 HTTP 요청을 메서드에 매핑할 수 있습니다. 
또한 `@RequestBody`, `@RequestParam`을 사용해 요청의 바디나 파라미터를 받을 수 있고, `@ResponseStatus`, `@ResponseBody`를 통해 HTTP 응답의 상태 코드나 본문을 설정할 수 있습니다.

그리고 `@RestController`의 경우, `@Controller`에 `@ResponseBody`가 결합된 형태로, View를 거치지 않고 메서드의 반환 값을 바로 HTTP 응답 본문으로 변환하여 JSON 등의 데이터 형태로 전달합니다.
이러한 방식으로 Spring MVC에 기반하여 RESTful API를 구축할 수도 있으며, 이때 RESTful API는 URI를 통해 자원을 식별하고, HTTP 메서드를 활용해 해당 자원에 대한 생성, 조회, 수정, 삭제 등의 작업을 수행하는 아키텍처 스타일이라고 할 수 있습니다.

jamie9504 replied:
잘 정리해주셨네요.
웹 애플리케이션 개발 경험이 없다고 체크해주셨는데, 그러면 이번 레벨에서 학습해야하는 분량이 남들보다 많다고 느껴질 수 있습니다.
그때마다 키워드를 나열해두고, 우선순위를 설정한 후 우선순위에 맞춰서 학습할 수 있는 분량이나 깊이를 조절해보는 것도 도움이 될 거라 생각합니다.

awrion3 replied:
오 네, 말씀하신 피드백을 염두에 두고 스프링을 학습해가도록 하겠습니다, 감사합니다! 😊

jamie9504 reviewed (changes requested):
안녕하세요, 리원. 리뷰어 제이미입니다.

코멘트 남겨두었습니다.

awrion3 committed:
docs: update feature list structure and content

awrion3 committed:
refactor: change method names and set relative path

awrion3 committed:
refactor: add boolean methods for comparison

- rename exception to exception message
- relocate row mapper to fields

jamie9504 commented on src/main/java/roomescape/controller/AdminReservationTimeController.java:
```diff
@@ -0,0 +1,40 @@
+import roomescape.service.ReservationTimeService;
+
+@RestController
+@RequestMapping("/times")
```
`ReservationTime`과 `times`라는 용어가 혼용되고 있군요!

awrion3 replied:
아 네, 해당 부분의 경우 현재로서는 예약 시간이라는 의미를 담는 것이 중요하다고 여겨져, `ReservationTime`을 유지해보도록 하겠습니다.

jamie9504 commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,104 @@
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    public Reservation readReservationById(final Long id) {
```
read, find, get, search 등 주로 쓰는 메서드 prefix들에 대해 알아보고,
각각 어떨 때 주로 쓰는지 찾아보거나 크루들과 이야기해보세요.

awrion3 replied:
음 네, 아래와 같이 정리해 볼 수 있었습니다.
- get은 존재가 확실한 데이터를 단순히 가져올 때 사용되며, 주로 Controller 계층에서 활용된다고 생각했습니다.
- find는 대상 데이터가 존재하지 않을 수도 있다는 전제하에 탐색하는 메서드로, Service 계층에서 쓰일 수 있습니다.
- read는 데이터를 읽어오는 동작을 의미하며, DB에서 데이터를 불러오는 데에 적합하다고 여겨집니다.
- search는 특정 조건에 따라 여러 결과를 찾는 작업에 적합하며, 따라서 Service나 Repository 계층에서 사용될 수 있습니다.

jamie9504 commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,104 @@
+        }
+    }
+
+    private void checkDuplicates(final Reservation reservation) {
```
check라는 메서드가 내부에서 Exception을 던질 것이라 예측할 수 있을까요?

awrion3 replied:
엇 네, 다시 생각해보니 해당 메서드명으로는 예측하기 어려울 수 있겠네요. 좀 더 검증한다는 의미를 담도록 리팩토링 했습니다!

jamie9504 commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,104 @@
+        List<Reservation> reservations = readAllReservations();
+
+        boolean isDuplicate = reservations.stream()
+                .anyMatch(reserve -> reserve.isSameDate(reservation) &&
+                        reserve.isSameTimeId(reservation)
```
`reserve.isSameDate(reservation) && reserve.isSameTimeId(reservation)`
이들은 하나의 메서드로 줄일 수도 있을 것 같아요.

awrion3 replied:
앗 네, 하나의 메서드로 줄일 수 있었네요! 해당 부분 리팩토링 했습니다!

jamie9504 commented on src/test/java/roomescape/controller/AdminReservationControllerTest.java:
```diff
@@ -0,0 +1,73 @@
+
+    @DisplayName("어드민 예약 내역 페이지 컨트롤러가 jdbc template을 의존하고 있는지 확인한다")
+    @Test
+    void checkAdminReservationControllerJdbcTemplate_NoDependency() {
```
해당 테스트를 통해 얻을 수 있는 것은 무엇이 있나요?
테스트를 할 때는 무엇을 얻고 싶은지와 유지보수성에 대해 항상 함께 생각해보면 좋습니다.

awrion3 replied:
아 네, 해당 테스트는 미션 진행 단계를 점검하기 위해 추가된 코드였습니다. 그리고 테스트와 관련하여 목적성과 유지보수성을 고려해야 한다는 점 유념하도록 하겠습니다!

jamie9504 reviewed (changes requested):
안녕하세요, 리원. 리뷰어 제이미입니다.

코멘트 남겨두었습니다.
다음 요청시 별도의 코멘트 없이 Approve & Merge 예정입니다.

awrion3 committed:
refactor: add value check and test case

- add null check for Reservation
- add same date and time id check for Reservation
- add equals and hashcode

awrion3 committed:
refactor: use validate instead of check duplicates

- change row mapper into constant
- rename controllers

jamie9504 reviewed (approved):
안녕하세요 리원. 리뷰어 제이미입니다.

최종 리뷰 기한 이후 요청이므로 별도의 코멘트 없이 Approve & Merge 합니다.
남은 2레벨 미션도 즐겁게 보내시기 바랍니다.
